File: src/go/problems/remove_stars_test.go
// https://leetcode.com/problems/removing-stars-from-a-string

package problems

import (
	"testing"

	"github.com/dobermanch/leetcode/core"
)

type RemoveStars struct{}

func TestRemoveStars(t *testing.T) {
	gen := core.TestSuite[RemoveStars]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param("leet**cod*e").Result("lecoe")
	}).Add(func(tc *core.TestCase) {
		tc.Param("erase*****").Result("")
	}).Run(t)
}

func (RemoveStars) Solution(s string) string {
	stack := make([]byte, len(s))
	index := 0
	for i := 0; i < len(s); i++ {
		if s[i] == '*' {
			index--
		} else {
			stack[index] = s[i]
			index++
		}
	}

	return string(stack[:index])
}

File: src/python/challenges/problems/remove_stars_test.py
#https://leetcode.com/problems/removing-stars-from-a-string

from core.problem_base import *

class CanJump(ProblemBase):
    def Solution(self, s: str) -> str:
        stack = []
        for ch in s:
            if ch == '*':
                stack.pop()
            else:
                stack.append(ch)
        
        return "".join(stack)

if __name__ == '__main__':
    TestGen(CanJump) \
        .Add(lambda tc: tc.Param("leet**cod*e").Result("lecoe")) \
        .Add(lambda tc: tc.Param("erase*****").Result("")) \
        .Run()
